### PR TITLE
refactors junos modules to support persistent socket connections

### PR DESCRIPTION
**Pull request by privateip**
* updates junos_netconf module
* updates junos_command module
* updates junos_config module
* updates _junos_template module
* adds junos_rpc module
* adds junos_user module

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure due to python 2.4 syntax errors:

```
2017-02-13 23:28:10 Compiling ./lib/ansible/module_utils/junos_compat.py ...
2017-02-13 23:28:10   File "./lib/ansible/module_utils/junos_compat.py", line 1
2017-02-13 23:28:10     2015 Peter Sprygada, <psprygada@ansible.com>
2017-02-13 23:28:10     ^
2017-02-13 23:28:10 SyntaxError: invalid syntax
2017-02-13 23:28:10 
2017-02-13 23:28:10 Compiling ./lib/ansible/modules/network/junos/junos_config.py ...
2017-02-13 23:28:10   File "./lib/ansible/modules/network/junos/junos_config.py", line 280
2017-02-13 23:28:10     with locked_config:
2017-02-13 23:28:10                      ^
2017-02-13 23:28:10 SyntaxError: invalid syntax
2017-02-13 23:28:10 
2017-02-13 23:28:10 Compiling ./lib/ansible/modules/network/junos/junos_netconf.py ...
2017-02-13 23:28:10   File "./lib/ansible/modules/network/junos/junos_netconf.py", line 153
2017-02-13 23:28:10     result = {'changed': False, warnings=warnings}
2017-02-13 23:28:10                                         ^
2017-02-13 23:28:10 SyntaxError: invalid syntax
```

File: lib/ansible/modules/network/junos/_junos_template.py [hunk 1]
```diff
@@ -125,42 +125,34 @@ def main():
         transport=dict(default='netconf', choices=['netconf'])
     )
 
-    module = NetworkModule(argument_spec=argument_spec,
+    module = AnsibleModule(argument_spec=argument_spec,
                            supports_check_mode=True)
 
+    confirm
```
Comment (Qalthos) [inline lib/ansible/modules/network/junos/_junos_template.py]: This looks unnecessary?

File: lib/ansible/modules/network/junos/junos_config.py [hunk 2]
```diff
@@ -231,91 +240,59 @@ def config_to_commands(config):
...
         candidate = candidate.split('\n')
 
-    kwargs = dict()
-    kwargs['comment'] = module.params['comment']
-    kwargs['confirm'] = module.params['confirm']
-    kwargs[module.params['update']] = True
-    kwargs['commit'] = not module.check_mode
-    kwargs['replace'] = module.params['replace']
+    kwargs = {
+        #'comment': module.params['comment'],
```
Comment (Qalthos) [inline lib/ansible/modules/network/junos/junos_config.py]: Should these be removed entirely?

Comment (privateip) [reply lib/ansible/modules/network/junos/junos_config.py]: argh... good catch, those need to be implemented

File: lib/ansible/modules/network/junos/_junos_template.py (hunk 1 above)
Comment (privateip) [reply lib/ansible/modules/network/junos/_junos_template.py]: hey whats that doing there ?!?!?!?!? .... new PR coming shortly

File: lib/ansible/module_utils/junos.py [hunk 3]
```diff
@@ -16,314 +16,203 @@
...
+    finally:
+        unlock_configuration(module)
+
+def get_diff(module):
+    reply = get_configuration(module, compare=True, format='text')
+    output = reply.xpath('//configuration-output')
+    if output:
+        return output[0].text
+
+def load(module, candidate, action='merge', commit=False, format='xml')
```
Comment (Qalthos) [inline lib/ansible/module_utils/junos.py]: Need a colon at the end here

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/junos/junos_user.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+"""
+from functools import partial
+
+from ncclient.xml_ import new_ele, sub_ele, to_xml
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.junos import load
+from ansible.module_utils.six import iteritems
+
+ROLES = ['operator', 'read-onYly', 'super-user', 'unauthorized']
```
Comment (Qalthos) [inline lib/ansible/modules/network/junos/junos_user.py]: `read-onYly` should probably be `read-only`

**Review by Qalthos — CHANGES_REQUESTED** [review]

**Pull request merged**